### DISCUSSION TITLE
Remove libraries that are not needed after static build

### DISCUSSION THREAD

**Discussion by kevcodez** [nuxt/nuxt#8515] [Questions (Nuxt 2)]
I'm using Nuxt to build static sites (target: static). I am getting markdown through an API and [marked](https://github.com/markedjs/marked) to convert the markdown to HTML. It only needs to be done *once* during build.
Thus, it would not be necessary to include the marked library in the final static build.

Is there a way to do avoid including the entire library if it's not used after initial render (during static build)?

Imagine this simple component to render the markdown using marked

```vue
<template>
  <div class="prose lg:prose-xl" v-html="markdownAsHtml" />
</template>

<script>
import marked from "marked";

export default {
  props: ["markdownData"],
  name: "markdown",
  computed: {
    markdownAsHtml() {
        return marked(markdownData);
    }
  }
};
</script>
```

**Comment by pi0** [comment]
Hi @kevcodez. Normally you may use [content module](https://content.nuxtjs.org/) for this but for manual approach you can lazy import it inside `fetch`

```vue
<template>
  <div class="prose lg:prose-xl" v-html="markdownAsHtml" />
</template>

<script>
import marked from "marked";

export default {
  props: ["markdownData"],
  name: "markdown",
  data() {
    this.markdownAsHtml = ''
  },
  async fetch() {
     const marked = await import('marked').them(m => m.default)
     this.markdownAsHtml = marked(this.markdownData)
  }
};
</script>
```

A more advanced version to also watch prop:

```vue
<template>
  <div class="prose lg:prose-xl" v-html="markdownAsHtml" />
</template>

<script>
import marked from "marked";

export default {
  props: ["markdownData"],
  name: "markdown",
  data() {
    this.markdownAsHtml = ''
  },
  async fetch() {
     await tyhis.renderMarkdown()
  },
  watch: {
    markdownData: 'renderMarkdown'
  },
  methods: {
    async renderMarkdown() {
      const marked = await import('marked').them(m => m.default)
     this.markdownAsHtml = marked(this.markdownData)
    }
  }
};
</script>
```

**Comment by kevcodez** [comment]
@pi0 Is it required to change the babel configuration?

I was getting the following exception:

> Error in fetch(): regeneratorRuntime is not defined    

FYI, I have a component library based on [vue-sfc-rollup](https://www.npmjs.com/package/vue-sfc-rollup). Including that component library in my Nuxt.js project.

**Reply by pi0** [reply]
@kevcodez I would suggest not transpiling `.vue` files from your component library. This way nuxt will take care of both `async` syntax and dynamic imports for creating proper chunks and also `fetch` is a nuxt specific feature...

-- -

BTW I didn't tried `vue-sfc-rollup` but it seems not properly transpiling `async/await` syntax. You may modify babel config or use `build.transpile` option in nuxt to include component library or simply avoid syntax like this:


```js
   fetch() {
      return import('marked').then(m => m.default).then(marked => { this.markdownAsHtml = marked(this.markdownData) })
  }
```

Also you can give [bili](https://github.com/egoist/bili) try. Still it won't fix dynamic chunk problem to remove marked from production bundle

**Comment by kevcodez** [comment]
@pi0 Do you know of any sample library I can have a look at?

The library of Vue components is only used internally and is specifically for Nuxt.js projects, btw.

**Reply by pi0** [reply]
Not a particular public library in mind but may i ask what is reason to bundle `.vue` SFCs in component lib?

**Reply by kevcodez** [reply]
I am building a site generator.

One Nuxt.js application is used to build the site visually with a preview, it will be a simple SPA, no static build. Basically a CMS.
One Nuxt.js application is used to statically generate the page based on the CMS data.

I want to have a common library for the components that can be used in both, the preview and the statically generated page.
For the statically generated page, I want to optimize the bundle size - meaning stripping of dependencies not needed after the static build.

Let me know if I need to elaborate further.

I'm open to alternative solutions if you have ideas to simplify this.

Theoretically, the library would not need any bundling at all since the nuxt.js application will do the entire bundling anyway.

**Comment by devlsh** [comment]
@kevcodez How do you deploy the static sites? It doesn't hugely matter how you do the inside CMS previews, since you seem to have that down with the runtime compilation, but if I could get more insight in to your deploy process maybe we could figure out how to include only production dependencies.
